lib/backends/sdc/containers.js

```diff
@@ -3578,6 +3578,14 @@ function _runExec(opts, callback) {
 
     serverSocket.on('connect', setupListeners);
 
+    // error can happen before connect too (eg. ECONNREFUSED)
+    serverSocket.on('error', function (error) {
+        opts.log.error('serverSocket for %s threw an error %s',
+            cmdString, error.toString());
+
+        endSocket(error);
```

Comment:
The reason is just that they're in different scopes. _runAttach will be run potentially multiple times (by multiple attaching clients) and the endSocket function there is within the scope of the setupListeners() function. It also has a comment:

```
        // When using multi attach the _endSocket callback will be GC'd
        // if defined outside of this scope
```

which made me think changing the scope wasn't a great idea. You're probably right that they should do the same thing though. Does changing this to not call endSocket and instead just doing:

```
    // error can happen before connect too (eg. ECONNREFUSED)
    serverSocket.on('error', function _onServerSocketError(error) {
        opts.log.error('serverSocket for %s threw an error %s',
            cmdString, error.toString());

        serverSocket.end();
        cb(error);
    });
```

make more sense?